### PR TITLE
Check also datatype sizes for floats to avoid losing bits

### PR DESCRIPTION
**Pull request by ohm314**
**Description**

HDF5 treats floats and doubles as Float, one datatype class. so reading a double dataset into a float vector won't even cause a warning message. This PR extends DataTypeClass a little bit to allow comparing DataTypeClasses with bit operators and then treats the case specially where datatypes are floating types and also checks the type size, printing a warning if there is a mismatch.

**How to test this?**

Here's a minimal patch to one of the examples that tests this:

```
--- a/src/examples/create_dataset_double.cpp
+++ b/src/examples/create_dataset_double.cpp
@@ -6,6 +6,7 @@
  *          http://www.boost.org/LICENSE_1_0.txt)
  *
  */
+#include <array>
 #include <iostream>
 #include <string>
 #include <vector>
@@ -37,6 +38,15 @@ int main(void) {
         // write it
         dataset.write(data);

+        std::vector<std::array<float, 6>> lore(2);
+        DataSet readback = file.getDataSet(DATASET_NAME);
+        readback.read(lore);
+        for (const auto& arr: lore) {
+            for (const auto e: arr) {
+                std::cout << " " << e;
+            }
+            std::cout << std::endl;
+        }
     } catch (Exception& err) {
         // catch and print any HDF5 error
         std::cerr << err.what() << std::endl;
```

```bash
cmake .. -DHIGHFIVE_EXAMPLES=ON
make -j8
src/examples/create_dataset_double_bin
```

### CODE REVIEW

**Review by matz-e — DISMISSED** [review]
Looks good _to me_

File: include/highfive/H5DataType.hpp [hunk 1]
```diff
@@ -37,6 +38,16 @@ enum class DataTypeClass {
...
 };
 
+inline DataTypeClass operator|(DataTypeClass lhs, DataTypeClass rhs) {
+    using T = std::underlying_type<DataTypeClass>::type;
+    return static_cast<DataTypeClass>(static_cast<T>(lhs) | static_cast<T>(rhs));
+}
+
+inline DataTypeClass operator&(DataTypeClass lhs, DataTypeClass rhs) {
+    using T = std::underlying_type<DataTypeClass>::type;
+    return static_cast<DataTypeClass>(static_cast<T>(lhs) & static_cast<T>(rhs));
```
Comment (1uc) [inline include/highfive/H5DataType.hpp:48]: This looks like bitfield manipulations to me, but when I look at `DataTypeClass` it's just an `enum class` without giving the individual values a particular value. In a small reproducer I get that they're numbered through: `0, 1, 2, 3, ..., n`. Therefore, I'm not sure I understand theses lines of code correctly.

Comment (ohm314) [reply include/highfive/H5DataType.hpp:48]: This is just some convenient trick to be able to check multiple instances of this enum with a specific value. Like this allows you to more elegantly write this

```c++
if ((dtype.getClass() & data_type.getClass()) == DataTypeClass::Float) {
    // ...
}
```

instead of this

```c++
if ((dtype.getClass() == DataTypeClass::Float) and (data_type.getClass() == DataTypeClass::Float)) {
    // ...
}
```

if it were enums instead of enum classes, they would implicitly convert to ints and you coudl do this without defining bitwise operators, but for enum classes, one needs to add the operator. Full disclosure, the trick comes (of course, where else) from here: https://softwareengineering.stackexchange.com/a/204566

Comment (ohm314) [reply include/highfive/H5DataType.hpp:48]: Let me rephrase: ~elegantly~ briefly
since not everyone will find that elegant.

Comment (1uc) [reply include/highfive/H5DataType.hpp:48]: By assigning powers of two as values, the bit logic should workout.

**Review by 1uc — APPROVED** [review]
